### ISSUE TITLE
Unexpected EOF for remote file

### ISSUE THREAD

**Issue by psp589** [robinrodricks/FluentFTP#308]
**FTP OS:** Windows 

**FTP Server:** Filezilla

**Computer OS:** windows 10 running an android emulator with Nougat 7.1

When I try to download a file from UWP it works without problems, when I try to download the same file using an android emulator I got an exception : 

FluentFTP.FtpException: Error while downloading the file from the server. See InnerException for more info. ---> System.IO.IOException: Unexpected EOF for remote file CAPT0004/db.zip [3744141/3744181 bytes read]
  at FluentFTP.FtpClient.DownloadFileInternal (System.String remotePath, System.IO.Stream outStream, System.IProgress`1[T] progress) [0x000c6] in <1bd16e2bef914c27b4aadfb191e1fd86>:0 
   --- End of inner exception stack trace ---
  at FluentFTP.FtpClient.DownloadFileInternal (System.String remotePath, System.IO.Stream outStream, System.IProgress`1[T] progress) [0x00283] in <1bd16e2bef914c27b4aadfb191e1fd86>:0 
  at FluentFTP.FtpClient.DownloadFileToFile (System.String localPath, System.String remotePath, System.Boolean overwrite, FluentFTP.FtpVerify verifyOptions, System.IProgress`1[T] progress) [0x00073] in <1bd16e2bef914c27b4aadfb191e1fd86>:0 
  at FluentFTP.FtpClient.DownloadFile (System.String localPath, System.String remotePath, System.Boolean overwrite, FluentFTP.FtpVerify verifyOptions, System.IProgress`1[T] progress) [0x0005b] in <1bd16e2bef914c27b4aadfb191e1fd86>:0 
  
I read something about setting the data type to Binary  so I execute this couple of lines before calling DownloadFile method : 
            _FtpClient.DownloadDataType = FtpDataType.Binary;
            _FtpClient.UploadDataType = FtpDataType.Binary;
But I get the same error 

**Logs :**

```
06-15 12:43:28.976 I/mono-stdout(24763): 
06-15 12:43:28.976 I/mono-stdout(24763): # FileExists("CAPT0004/db.zip")
06-15 12:43:28.976 I/mono-stdout(24763): 
06-15 12:43:28.977 I/mono-stdout(24763): # GetWorkingDirectory()
06-15 12:43:28.977 I/mono-stdout(24763): Command:  PWD
Response: 257 "/" is current directory.
Command:  SIZE /CAPT0004/db.zip
06-15 12:43:28.986 I/mono-stdout(24763): Response: 257 "/" is current directory.
06-15 12:43:28.986 I/mono-stdout(24763): Command:  SIZE /CAPT0004/db.zip
Response: 213 374421006-15 12:43:28.990 I/mono-stdout(24763): Response: 213 3744210


# DownloadFile("storage/emulated/0/Android/data/com.companyname.Preventa.View/files/.__override__/db.zip", "CAPT0004/db.zip", True, Retry, Delete, Throw)

# GetFileSize("CAPT0004/db.zip")
Command:  SIZE CAPT0004/db.zip06-15 12:43:31.515 I/mono-stdout(24763): 

06-15 12:43:31.515 I/mono-stdout(24763): # DownloadFile("storage/emulated/0/Android/data/com.companyname.Preventa.View/files/.__override__/db.zip", "CAPT0004/db.zip", True, Retry, Delete, Throw)
06-15 12:43:31.516 I/mono-stdout(24763): 
06-15 12:43:31.517 I/mono-stdout(24763): # GetFileSize("CAPT0004/db.zip")
06-15 12:43:31.517 I/mono-stdout(24763): Command:  SIZE CAPT0004/db.zip
06-15 12:43:31.522 I/mono-stdout(24763): Response: 213 3744210Response: 213 3744210

# OpenRead("CAPT0004/db.zip", Binary, 0)

# GetFileSize("CAPT0004/db.zip")
Command:  SIZE CAPT0004/db.zip

06-15 12:43:31.522 I/mono-stdout(24763): 
06-15 12:43:31.522 I/mono-stdout(24763): # OpenRead("CAPT0004/db.zip", Binary, 0)
06-15 12:43:31.522 I/mono-stdout(24763): 
06-15 12:43:31.522 I/mono-stdout(24763): # GetFileSize("CAPT0004/db.zip")
06-15 12:43:31.522 I/mono-stdout(24763): Command:  SIZE CAPT0004/db.zip
06-15 12:43:31.526 I/mono-stdout(24763): Response: 213 3744210Response: 213 3744210

# OpenPassiveDataStream(AutoPassive, "RETR CAPT0004/db.zip", 0)
Command:  EPSV

06-15 12:43:31.526 I/mono-stdout(24763): 
06-15 12:43:31.526 I/mono-stdout(24763): # OpenPassiveDataStream(AutoPassive, "RETR CAPT0004/db.zip", 0)
06-15 12:43:31.526 I/mono-stdout(24763): Command:  EPSV
06-15 12:43:31.542 I/art     (24763): Starting a blocking GC Explicit
06-15 12:43:31.550 I/art     (24763): Explicit concurrent mark sweep GC freed 2213(103KB) AllocSpace objects, 0(0B) LOS objects, 31% free, 8MB/12MB, paused 568us total 7.696ms
06-15 12:43:31.551 D/Mono    (24763): GC_TAR_BRIDGE bridges 205 objects 8402 opaque 4004 colors 205 colors-bridged 205 colors-visible 205 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.04ms tarjan 1.40ms scc-setup 0.04ms gather-xref 0.01ms xref-setup 0.01ms cleanup 0.07msResponse: 229 Entering Extended Passive Mode (|||5521|)
Status:   Connecting to ***:5521
Command:  RETR CAPT0004/db.zip

06-15 12:43:31.551 D/Mono    (24763): GC_BRIDGE: Complete, was running for 9.01ms
06-15 12:43:31.551 D/Mono    (24763): GC_MINOR: (Nursery full) time 6.46ms, stw 7.24ms promoted 819K major size: 9712K in use: 8923K los size: 3072K in use: 1996K
06-15 12:43:31.553 I/mono-stdout(24763): Response: 229 Entering Extended Passive Mode (|||5521|)
06-15 12:43:31.554 I/mono-stdout(24763): Status:   Connecting to ***:5521
06-15 12:43:31.556 I/mono-stdout(24763): Command:  RETR CAPT0004/db.zip
Response: 150 Connection accepted
06-15 12:43:31.572 I/mono-stdout(24763): Response: 150 Connection accepted
Status:   Disposing FtpSocketStream...06-15 12:43:31.913 I/mono-stdout(24763): Status:   Disposing FtpSocketStream...

06-15 12:43:34.885 I/Choreographer(24763): Skipped 181 frames!  The application may be doing too much work on its main thread.Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   The stale data was: 226 Transfer OK
Status:   Disposing FtpSocketStream...

06-15 12:43:34.889 I/mono-stdout(24763): Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
06-15 12:43:34.893 I/mono-stdout(24763): Status:   The stale data was: 226 Transfer OK
Status:   Not sending QUIT because the connection has already been closed.
Status:   Disposing FtpSocketStream...

# Dispose()
Status:   Disposing FtpClient object...
Status:   Disposing FtpSocketStream...

```
I'm new to linux / android, there is some bug? Am I doing something wrong? some configuration needed? 
Thank you in advance.

**Comment by jersiovic** [comment]
Same problem here

**Comment by jersiovic** [comment]
It looks problem is related to this problem in Android: https://wiki.sei.cmu.edu/confluence/display/java/FIO10-J.+Ensure+the+array+is+filled+when+using+read%28%29+to+fill+an+array

**Comment by jersiovic** [comment]
It looks problems were solved replacing this line https://github.com/robinrodricks/FluentFTP/blob/master/FluentFTP/Client/FtpClient_HighLevel.cs#L1795

Per this other:

´´´
while (((readBytes = downStream.Read(buffer, 0, buffer.Length)) > 0) || 
                                    (readBytes < 0 && downStream.Length>0 && downStream.Position < downStream.Length))
´´´

**Comment by jersiovic** [comment]
A small correction to the condition. Instead of `readbytes <0` it should be `readbytes <=0`.
I mean this:
```
while (((readBytes = downStream.Read(buffer, 0, buffer.Length)) > 0) || 
                                    (readBytes <= 0 && downStream.Length>0 && downStream.Position < downStream.Length))
```

**Comment by jersiovic** [comment]
I have tested it on Android and on UWP if u want I can send a PR

**Comment by jersiovic** [comment]
Sorry but after a few days working properly, now I still have same problem. So, my fix seems that did't solve the problem.

**Comment by robinrodricks** [comment]
Released in [FluentFTP 19.2.3](https://www.nuget.org/packages/FluentFTP/19.2.3). 

Please check if it works.

**Comment by robinrodricks** [comment]
Please re-open if the issue still persists.
